### PR TITLE
PHOENIX-5389

### PR DESCRIPTION
**Pull request by comnetwork**

### CODE REVIEW

**Comment by comnetwork** [comment]
This PR is mainly for:
1. Add a new JoinTable.createSubJoinTable method to replace the old JoinTable.getSubJoinTableWithoutPostFilters,  JoinTable.createSubJoinTable also push down the postFilter to the new created JoinTable in addition to create the new JoinTable.
2. Add a JoinCompiler.PushDownPostFilterParseNodeVisitor as the visitor to  push down the postFilter.
3. When add JoinCompiler.PushDownPostFilterParseNodeVisitor, two abstract class AndRewriterBooleanParseNodeVisitor and AndBooleanParseNodeVisitor are added to eliminate the duplicate code with other  BooleanParseNodeVisitor(SubqueryRewriter.JoinConditionExtractor, QueryOptimizer.WhereConditionRewriter etc.)
4. Refactor JoinCompiler.WhereNodeVisitor/JoinTable and JoinCompiler.OnNodeVisitor/JoinSpec to make the modification internal states of JoinTable and JoinSpec sealed in  JoinTable and JoinSpec class.

File: phoenix-core/src/main/java/org/apache/phoenix/compile/JoinCompiler.java [hunk 1]
```diff
@@ -438,9 +471,46 @@ public Expression compilePostFilterExpression(StatementContext context) throws S
             return vector;
         }
 
-        public JoinTable getSubJoinTableWithoutPostFilters() {
-            return joinSpecs.size() > 1 ? new JoinTable(table, joinSpecs.subList(0, joinSpecs.size() - 1)) :
-                new JoinTable(table);
+        /**
```
Comment (comnetwork) [inline phoenix-core/src/main/java/org/apache/phoenix/compile/JoinCompiler.java]: Added a new JoinTable.createSubJoinTable method to replace the old JoinTable.getSubJoinTableWithoutPostFilters.
In addition to create the new JoinTable, JoinTable.createSubJoinTable also pushs down the PostFilter to the new created JoinTable .

File: phoenix-core/src/main/java/org/apache/phoenix/compile/JoinCompiler.java [hunk 2]
```diff
@@ -517,8 +597,24 @@ public JoinType getType() {
             return onConditions;
         }
 
-        public JoinTable getJoinTable() {
-            return joinTable;
+        public JoinTable getRhsJoinTable() {
```
Comment (comnetwork) [inline phoenix-core/src/main/java/org/apache/phoenix/compile/JoinCompiler.java]: add many methods in JoinSpec for OnNodeVisitor to make the modification of internal states of  JoinSpec is sealed in JoinSpec class.

File: phoenix-core/src/main/java/org/apache/phoenix/compile/JoinCompiler.java [hunk 3]
```diff
@@ -320,32 +328,57 @@ public ParseNode getPostFiltersCombined() {
             return combine(postFilters);
         }
 
-        public void addFilter(ParseNode filter) throws SQLException {
+        public void addPostJoinFilter(ParseNode parseNode) {
```
Comment (comnetwork) [inline phoenix-core/src/main/java/org/apache/phoenix/compile/JoinCompiler.java]: add many methods in JoinTable for JoinCompiler.WhereNodeVisitor to make the modification of internal states of JoinTable is sealed in JoinTable class.

File: phoenix-core/src/main/java/org/apache/phoenix/compile/JoinCompiler.java [hunk 4]
```diff
@@ -320,32 +328,57 @@ public ParseNode getPostFiltersCombined() {
...
+        public List<JoinSpec> getPrefilterAcceptedJoinSpecs() {
+            return this.prefilterAcceptedTables;
+        }
+
+        /**
+         * try to decompose filter and push down to single table.
+         * @param filter
+         * @throws SQLException
+         */
+        public void pushDwonFilter(ParseNode filter) throws SQLException {
```
Comment (twdsilva) [inline phoenix-core/src/main/java/org/apache/phoenix/compile/JoinCompiler.java]: nit: typo rename to pushDownFilter

File: phoenix-core/src/test/java/org/apache/phoenix/compile/QueryCompilerTest.java [hunk 5]
```diff
@@ -5978,4 +5979,210 @@ public void testDistinctCountLimitBug5217() throws Exception {
...
+        queryPlan = TestUtil.getOptimizeQueryPlanNoIterator(conn, sql);
+        hashJoinPlan = (HashJoinPlan)queryPlan;
+        assertTrue(hashJoinPlan.getJoinInfo().getPostJoinFilterExpression().toString().equals(
+                "(I.SUPPLIER_ID != 'medi' OR S.ADDRESS = 'hai' OR O.QUANTITY = 8)"));
+        hashSubPlans = (HashSubPlan[])hashJoinPlan.getSubPlans();
+        assertTrue(hashSubPlans.length == 1);
+        subHashJoinPlan = (HashJoinPlan)(hashSubPlans[0].getInnerPlan());
+        assertTrue(subHashJoinPlan.getJoinInfo().getPostJoinFilterExpression() == null);
+
+        //one condition can not pushdwon and two condition can pushdown.
```
Comment (twdsilva) [inline phoenix-core/src/test/java/org/apache/phoenix/compile/QueryCompilerTest.java]: nit: typo pushDown

**Comment by comnetwork** [comment]
@twdsilva , thank you for the review, I would make a modification.

**Comment by gjacoby126** [comment]
The associated JIRA is already resolved -- closing PR

**Pull request closed without merging**